core/typedefs.h

```diff
@@ -92,7 +92,7 @@
 #endif
 
 #ifndef SGN
-#define SGN(m_v) (((m_v) < 0) ? (-1.0) : (+1.0))
+#define SGN(m_v) (((m_v) == 0) ? (m_v) : (((m_v) < 0) ? (-1.0) : (+1.0)))
```

Comment:
May as well just return zero:

```suggestion
#define SGN(m_v) (((m_v) == 0) ? (0) : (((m_v) < 0) ? (-1.0) : (+1.0)))
```

Reply:
I hadn't noticed that this bug was already fixed in GDScript's implementation of `sign()`, that's good to know.  I found it while trying to use `Vector2::sign()` which led me to `SGN`.

I was on the fence about returning 0, but I went with (m_v) to reduce the chance of implicit type conversion.  I'm not sure if it would actually be an issue but I'd rather avoid it.  Returning a literal value for +/- 1.0 is unavoidable, however.  The only case I can think of where returning a literal would be preferable is if (m_v) is an expression with side-effects; thus embedding it multiple times would repeat the mutation.  But since there is no way to write this tripartite branch without invoking (m_v) at least *twice* already, I figured invoking it a third time shouldn't be an issue, since (m_v) already mustn't have side effects.

Since the bug doesn't actually exist in the GDScript implementation, even though it's repeating code I'd rather not fix what's not broken, personally.

Reply:
Well IMO it should be `0.0`. it doesn't have to be the same type as `m_v`, it has to be the same time as `-1.0` and `+1.0` which are the other two possible values.

Reply:
0.0 does make more sense now that I think about it; I have amended the commit to reflect that.

Reply:
Although changing `-1.0` to `-1` etc would also make sense, if you need to store the result in an integer.

Reply:
This is probably float by design, I don't see much reason to change it.